File: app/src/main/java/com/absinthe/kage/connect/tcp/Request.java
package com.absinthe.kage.connect.tcp;

import android.util.Log;

import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.TimeUnit;

public class Request extends Packet {
    private static final String TAG = Request.class.getSimpleName();

    private ArrayBlockingQueue<Response> responses = new ArrayBlockingQueue<>(1);
    private String id;

    public void setResponse(Response response) {
        try {
            responses.put(response);
        } catch (InterruptedException e) {
            Log.e(TAG, e.toString());
        }
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public Response waitResponse(int timeout) {
        Response response = null;
        try {
            response = responses.poll(timeout, TimeUnit.MILLISECONDS);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        return response;
    }
}

File: app/src/main/kotlin/com/absinthe/kage/connect/tcp/Request.kt
package com.absinthe.kage.connect.tcp

import com.absinthe.kage.device.heartbeat.ErrorResponse
import timber.log.Timber
import java.util.concurrent.ArrayBlockingQueue
import java.util.concurrent.TimeUnit

open class Request : Packet() {

    var id: String = System.currentTimeMillis().toString()
    private val responses = ArrayBlockingQueue<Response>(1)

    fun setResponse(response: Response) {
        try {
            responses.put(response)
        } catch (e: InterruptedException) {
            Timber.e(e)
        }
    }

    fun waitResponse(timeout: Int): Response {
        return try {
            responses.poll(timeout.toLong(), TimeUnit.MILLISECONDS)
        } catch (e: InterruptedException) {
            e.printStackTrace()
            ErrorResponse()
        }
    }
}